Libraries/oneMKL/fourier_correlation/fcorr_2d_usm.cpp

```diff
@@ -6,151 +6,299 @@
+    throw std::invalid_argument("The number of rows of the images, chosen as "
+                          "first input of the program, must be 6 or greater.");
+  const unsigned n_rows = temp;
+  temp = (argc <= 2) ? 8 : std::stoi(argv[2]);
+  // n_cols >= 7 required for the arbitrary signals as defined herein
+  if (temp < 7)
+    throw std::invalid_argument("The number of columns of the images, chosen as "
+                          "second input of the program, must be 7 or greater.");
+  const unsigned n_cols = temp;
+  const unsigned num_elem = n_rows * n_cols;
 
-    // Data layout in conjugate-even domain
-    std::int64_t complex_layout[4] = {0, c_stride_h, 1};
-    forward_plan.set_value(oneapi::mkl::dft::config_param::OUTPUT_STRIDES,
-                           complex_layout);
+  // Let s and p be integer s.t. 0 <= s < n_rows, 0 <= p < n_cols, and let
+  // corr(s, p) =
+  //  \sum_{j = 0}^{n_rows - 1} \sum_{k = 0}^{n_cols - 1} \
+  //    img1(j, k) * img2((j - s + n_rows) mod n_rows, (k - p + n_cols) mod n_cols)
+  // be the cross-correlation between two real periodic signals img1 and img2
+  // of periods (n_rows, n_cols). This code shows how to calculate corr using
+  // Discrete Fourier Transforms (DFTs).
+  // Note: in the above, the notation "x(i, j)" represents the entry of
+  // multi-index (i, j) within a (real) data sequence "x". If the latter is
+  // stored in memory in an allocation "x_alloc" using a unit row stride and a
+  // column stride col_stride (col_stride >= n_cols), we have
+  //           x_alloc[i * col_stride + j] <- "x(i,j)"
 
-    // Perform forward transforms on real arrays
-    forward_plan.commit(Q);
-    auto evt1 = oneapi::mkl::dft::compute_forward(forward_plan, img1);
-    auto evt2 = oneapi::mkl::dft::compute_forward(forward_plan, img2);
+  // This program returns 0 (resp. 1) if naive and DFT-based calculations are
+  // (resp. are not) within error tolerance of one another.
+  int return_code = 0;
+  
+  // Initialize SYCL queue
+  sycl::queue Q(sycl::default_selector_v);
+  std::cout << "Running on: "
+            << Q.get_device().get_info<sycl::info::device::name>()
+            << std::endl;
 
-    // Compute: DFT(img1) * CONJG(DFT(img2))
-    oneapi::mkl::vm::mulbyconj(Q, n_rows * c_stride_h,
-                               reinterpret_cast<std::complex<float>*>(img1),
-                               reinterpret_cast<std::complex<float>*>(img2),
-                               reinterpret_cast<std::complex<float>*>(corr),
-                               {evt1, evt2})
-                               .wait();
+  // Initialize 2D image and correlation arrays. The arrays must be large enough
+  // to store the forward and backward domains' data, consisting of
+  // n_rows * n_cols real values and n_rows * (n_cols / 2 + 1) complex values,
+  // respectively (for the DFT-based calculations).
+  // Note: 2 * (n_cols / 2 + 1) > n_cols for all n_cols > 0, so
+  //    max(n_rows * n_cols real,
```

Comment:
```suggestion
  //    max(n_rows * n_cols,
```
Minor, looks like an extra real here?

Reply:
Thank you, addressed in https://github.com/oneapi-src/oneAPI-samples/pull/2520/commits/595e1a6f749343a1a66d963c6aba495162cf510b